src/System.Memory/tests/ArrayBufferWriter/ArrayBufferWriterTests.Byte.cs

```diff
@@ -0,0 +1,367 @@
+                output.Advance(output.FreeCapacity);
+                Assert.Equal(capacity, output.WrittenCount);
+                Assert.Equal(0, output.FreeCapacity);
+            }
+
+            {
+                var output = new ArrayBufferWriter<byte>();
+                output.Advance(output.Capacity);
+                Assert.Equal(output.Capacity, output.WrittenCount);
+                Assert.Equal(0, output.FreeCapacity);
+                int previousCapacity = output.Capacity;
+                Span<byte> _ = output.GetSpan();
+                Assert.True(output.Capacity > previousCapacity);
+            }
+
+            {
+                var output = new ArrayBufferWriter<byte>();
+                WriteData(output, 2);
+                ReadOnlyMemory<byte> previousMemory = output.WrittenMemory;
+                ReadOnlySpan<byte> previousSpan = output.WrittenSpan;
+                Assert.True(previousSpan.SequenceEqual(previousMemory.Span));
+                output.Advance(10);
+                Assert.False(previousMemory.Span.SequenceEqual(output.WrittenMemory.Span));
+                Assert.False(previousSpan.SequenceEqual(output.WrittenSpan));
+                Assert.True(output.WrittenSpan.SequenceEqual(output.WrittenMemory.Span));
+            }
+        }
+
+        [Fact]
+        public static void AdvanceZero()
+        {
+            var output = new ArrayBufferWriter<byte>();
+            WriteData(output, 2);
+            Assert.Equal(2, output.WrittenCount);
+            ReadOnlyMemory<byte> previousMemory = output.WrittenMemory;
+            ReadOnlySpan<byte> previousSpan = output.WrittenSpan;
+            Assert.True(previousSpan.SequenceEqual(previousMemory.Span));
+            output.Advance(0);
+            Assert.Equal(2, output.WrittenCount);
+            Assert.True(previousMemory.Span.SequenceEqual(output.WrittenMemory.Span));
+            Assert.True(previousSpan.SequenceEqual(output.WrittenSpan));
+            Assert.True(output.WrittenSpan.SequenceEqual(output.WrittenMemory.Span));
+        }
+
+        [Fact]
+        public static void InvalidAdvance()
+        {
+            {
+                var output = new ArrayBufferWriter<byte>();
+                Assert.Throws<ArgumentException>(() => output.Advance(-1));
+                Assert.Throws<InvalidOperationException>(() => output.Advance(output.Capacity + 1));
+            }
+
+            {
+                var output = new ArrayBufferWriter<byte>();
+                WriteData(output, 100);
+                Assert.Throws<InvalidOperationException>(() => output.Advance(output.FreeCapacity + 1));
+            }
+        }
+
+        public static bool IsX64 { get; } = IntPtr.Size >= 8;
+
+        [ConditionalFact(nameof(IsX64))]
+        [OuterLoop]
+        public static void InvalidAdvance_Large()
+        {
+            try
+            {
+                {
+                    var output = new ArrayBufferWriter<byte>(2_000_000_000);
```

Comment:
Are these types of tests always going to succeed? It doesn't seem like a good idea to be allocating this much memory (even temporary) since some environments such as a local machine may not be able to succeed. Large object heap concerns too.

Reply:
This test only runs on outerloop and hence shouldn't affect dev box testing. We have such tests elsewhere as well.

This test validates the boundary condition of going over `int32.MaxValue`.

> Large object heap concerns too.

This is a test, why is that a concern?

Reply:
I didn't see these were outer loop tests. Large object heap comment was that the memory would be reserved until the process is closed.